weaviate/collection/grpc.py

```diff
@@ -434,20 +440,119 @@ def __extract_metadata_for_object(
+            return None
+
+    def _extract_metadata_for_object(
+        self,
+        add_props: "weaviate_pb2.ResultAdditionalProps",
+    ) -> _MetadataReturn:
+        return _MetadataReturn(
+            uuid=uuid_lib.UUID(add_props.id) if len(add_props.id) > 0 else None,
+            vector=[float(num) for num in add_props.vector] if len(add_props.vector) > 0 else None,
+            distance=add_props.distance if add_props.distance_present else None,
+            certainty=add_props.certainty if add_props.certainty_present else None,
+            creation_time_unix=add_props.creation_time_unix
+            if add_props.creation_time_unix_present
+            else None,
+            last_update_time_unix=add_props.last_update_time_unix
+            if add_props.last_update_time_unix_present
+            else None,
+            score=add_props.score if add_props.score_present else None,
+            explain_score=add_props.explain_score if add_props.explain_score_present else None,
+        )
+
+    def _deserialize_primitive(self, value: Any, type_value: Any) -> Any:
+        if type_value == uuid_lib.UUID:
+            return uuid_lib.UUID(value)
+        if type_value == datetime.datetime:
+            return datetime.datetime.fromisoformat(value)
+        if isinstance(type_value, list):
+            return [
+                self._deserialize_primitive(val, type_value[idx]) for idx, val in enumerate(value)
+            ]
+        return value
+
+
+class _GrpcCollection(_Grpc):
+    def __init__(self, connection: Connection, name: str, tenant: Optional[str]):
+        super().__init__(connection, name, tenant)
+
+    def __parse_result(
+        self, properties: "weaviate_pb2.ResultProperties", type_: Optional[Type[Properties]]
+    ) -> Properties:
+        hints = get_type_hints(type_) if get_origin(type_) is not dict and type_ is not None else {}
+
+        result = {}
+
+        for name, non_ref_prop in properties.non_ref_properties.items():
+            result[name] = self._deserialize_primitive(non_ref_prop, hints.get(name))
+
+        for number_array_property in properties.number_array_properties:
+            result[number_array_property.key] = [float(val) for val in number_array_property.vals]
+
+        for int_array_property in properties.int_array_properties:
+            result[int_array_property.key] = [int(val) for val in int_array_property.vals]
+
+        for text_array_property in properties.text_array_properties:
+            result[text_array_property.key] = [str(val) for val in text_array_property.vals]
+
+        for boolean_array_property in properties.boolean_array_properties:
+            result[boolean_array_property.key] = [bool(val) for val in boolean_array_property.vals]
+
+        for uuid_array_property in properties.uuid_array_properties:
+            result[uuid_array_property.key] = [
+                uuid_lib.UUID(val) for val in uuid_array_property.vals
+            ]
```

Comment:
is the code from where this is copied not used anymore? If not, can you delete it?

Reply:
But why can't this be shared between collections and models? This is not type-relevant and would be the same code for both?

Reply:
There is potential for the code to be shared between collections and models if the generics play nicely between `Properties` and `Model` but I'm unsure of this at this moment. I think it will be shareable once we have a better idea of the ORM capabilities, can it wait until then?

But the function itself depends on `type_: Optional[Type[Properties]]`, which causes the potential issue. The loops through the properties could be shared in some function that takes `result` as an argument and mutates it, but I'm not personally fond of passing variables by reference and mutating them within functions